druid/src/widget/textbox.rs

```diff
@@ -435,13 +435,7 @@ impl Widget<String> for TextBox {
 
     fn paint(&mut self, ctx: &mut PaintCtx, data: &String, env: &Env) {
         // Guard against changes in data following `event`
-        let content = if data.is_empty() {
-            &self.placeholder
-        } else {
-            data
-        };
-
-        self.selection = self.selection.constrain_to(content);
+        self.selection = self.selection.constrain_to(data);
```

Comment:
I think this means that if you have a selection, then your text gets disappeared by an update, it will wind up not drawing any selection as the selection will also disappear. It might be clearer to explicitly not do any of the selection stuff if we are using the placeholder - not sure if the same applies for the cursor?

Reply:
this is a questionable line, and maybe it should be removed? As I understand it it's an attempt to catch some problem with an invalid selection that was not correctly handled during update.

The problem is that the selection is stored in the widget, but it references the app data, so when the app data is changed in another widget, the selection can end up invalid. This *should* be handled in `update`, but maybe there was some problem?

Reply:
Yes it does seem a bit odd

Reply:
Maybe this could be an assert instead of trying to silently fix it? It seems to me like a serious bug in druid if we get a paint call with different data from our last update...

Reply:
I'm not quite sure how we could land in paint without having hit update to constrain the selection either. That definitely sounds like an odd case. Do we want to drop this line and see if we have crashes again, or would we rather leave this as a backstop? I don't see how this could be terribly expensive.

Reply:
Feels to me this line should be removed, just on the basis that paint really should be as close to a pure function as possible. If there is a user-visible problem here (and I'm not sure there is, but haven't done detailed code analysis or experimentation), it should probably be handled in update.

Reply:
I'm in the process of rewriting this stuff, and it will be obsoleted soon. My inclination is to just leave it as is for now?